lib/active_storage/service/cloudinary_service.rb

```diff
@@ -0,0 +1,156 @@
+require 'cloudinary'
+require 'open-uri'
+
+module ActiveStorage
+  class Service::CloudinaryService < Service
+    # FIXME: implement setup for private resource type
+    # FIXME: allow configuration via cloudinary url
+    def initialize(cloud_name:, api_key:, api_secret:, options: {})
+      options.merge!(
+        cloud_name: cloud_name,
+        api_key: api_key,
+        api_secret: api_secret
+      )
+      Cloudinary.config(options)
+      # Cloudinary.config_from_url(url)
+    end
+
+    def upload(key, io, checksum: nil)
+      instrument :upload, key: key, checksum: checksum do
+        Cloudinary::Uploader.upload(io, public_id: key)
+      end
+    end
+
+    # Return the content of the file at the +key+.
+    def download(key)
+      tmp_file = open(url_for_public_id(key))
+      if block_given?
+        instrument :streaming_download, key: key do
+          File.open(tmp_file, 'rb') do |file|
+            while (data = file.read(64.kilobytes))
+              yield data
+            end
+          end
+        end
+      else
+        instrument :download, key: key do
+          File.binread tmp_file
+        end
+      end
+    end
+
+    # Delete the file at the +key+.
+    def delete(key)
+      instrument :delete, key: key do
+        delete_resource_with_public_id(key)
+      end
+    end
+
+    # Delete files at keys starting with the +prefix+.
+    def delete_prefixed(prefix)
+      instrument :delete_prefixed, prefix: prefix do
+        find_resources_with_public_id_prefix(prefix).each do |resource|
+          delete_resource_with_public_id(resource['public_id'])
+        end
+      end
+    end
+
+    # Return +true+ if a file exists at the +key+.
+    def exist?(key)
+      instrument :exist?, key: key do
+        resource_exists_with_public_id?(key)
+      end
+    end
+
+    # Returns a signed, temporary URL for the file at the +key+. The URL will be valid for the amount
+    # of seconds specified in +expires_in+. You most also provide the +disposition+ (+:inline+ or +:attachment+),
+    # +filename+, and +content_type+ that you wish the file to be served with on request.
+    def url(key, expires_in:, disposition:, filename:, content_type:)
```

Comment:
is it more flexible to have a default for some of the required arguments?

Reply:
Yea, probably. But since this is the required [activestorage service interface](https://github.com/rails/rails/blob/master/activestorage/lib/active_storage/service.rb) I figured it should break if the interface inadvertently changes